sys/crypto/poly1305.c

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (C) 2016 Andrew Moon (dedicated to the public domain)
+ * Copyright (C) 2018 Freie Universität Berlin
+ * Copyright (C) 2018 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+#include "crypto/poly1305.h"
+
+static void poly1305_blocks(poly1305_ctx_t *ctx, const uint8_t *msg, size_t msglen);
+static uint32_t u8to32(const uint8_t *p)
+{
+    return
+        (((uint32_t)(p[0] & 0xff)) |
+         ((uint32_t)(p[1] & 0xff) <<  8) |
+         ((uint32_t)(p[2] & 0xff) << 16) |
+         ((uint32_t)(p[3] & 0xff) << 24));
+}
+
+static void u32to8(uint8_t *p, uint32_t v)
+{
+    p[0] = (v) & 0xff;
+    p[1] = (v >>  8) & 0xff;
+    p[2] = (v >> 16) & 0xff;
+    p[3] = (v >> 24) & 0xff;
+}
+
+void poly1305_update(poly1305_ctx_t *ctx, const uint8_t *data, size_t len)
+{
+    size_t i;
+
+    /* handle leftover */
+    if (ctx->leftover) {
+        size_t want = (poly1305_block_size - ctx->leftover);
+        if (want > len) {
+            want = len;
+        }
+        for (i = 0; i < want; i++) {
+            ctx->buffer[ctx->leftover + i] = data[i];
+        }
+        len -= want;
+        data += want;
+        ctx->leftover += want;
+        if (ctx->leftover < poly1305_block_size) {
+            return;
+        }
+        poly1305_blocks(ctx, ctx->buffer, poly1305_block_size);
+        ctx->leftover = 0;
+    }
+
+    /* process full blocks */
+    if (len >= poly1305_block_size) {
+        size_t want = (len & ~(poly1305_block_size - 1));
+        poly1305_blocks(ctx, data, want);
+        data += want;
+        len -= want;
+    }
+
+    /* store leftover */
+    if (len) {
+        for (i = 0; i < len; i++)
+            ctx->buffer[ctx->leftover + i] = data[i];
+        ctx->leftover += len;
+    }
+}
+
+void poly1305_init(poly1305_ctx_t *ctx, const uint8_t *key)
+{
+    /* r &= 0xffffffc0ffffffc0ffffffc0fffffff */
+    ctx->r[0] = (u8to32(&key[ 0])) & 0x3ffffff;
+    ctx->r[1] = (u8to32(&key[ 3]) >> 2) & 0x3ffff03;
+    ctx->r[2] = (u8to32(&key[ 6]) >> 4) & 0x3ffc0ff;
+    ctx->r[3] = (u8to32(&key[ 9]) >> 6) & 0x3f03fff;
+    ctx->r[4] = (u8to32(&key[12]) >> 8) & 0x00fffff;
+
+    /* h = 0 */
+    ctx->h[0] = 0;
+    ctx->h[1] = 0;
+    ctx->h[2] = 0;
+    ctx->h[3] = 0;
+    ctx->h[4] = 0;
+
+    /* save pad for later */
+    ctx->pad[0] = u8to32(&key[16]);
+    ctx->pad[1] = u8to32(&key[20]);
+    ctx->pad[2] = u8to32(&key[24]);
+    ctx->pad[3] = u8to32(&key[28]);
+
+    ctx->leftover = 0;
+    ctx->final = 0;
+}
+
+static void poly1305_blocks(poly1305_ctx_t *ctx, const uint8_t *data, size_t len)
+{
+    const uint32_t hibit = (ctx->final) ? 0 : (1UL << 24); /* 1 << 128 */
+    uint32_t r0, r1, r2, r3, r4;
+    uint32_t s1, s2, s3, s4;
+    uint32_t h0, h1, h2, h3, h4;
+    uint64_t d0, d1, d2, d3, d4;
```

Comment:
please move the d and c variables inside the scope of the while loop so the CI is happy.